### PR TITLE
Add `tensor_split` function, based on `numpy.array_split`

### LINKED ISSUE #9382: tensor.chunk returns wrong number of chunks
## Issue description
Number of chunks generated by `torch.chunk` doesn't match `chunks` parameter.

## Code example
```
>>> x = torch.arange(0, 11)
>>> x.chunk(5)
(tensor([ 0.,  1.,  2.]),
 tensor([ 3.,  4.,  5.]),
 tensor([ 6.,  7.,  8.]),
 tensor([  9.,  10.]))
>>> len(x.chunk(chunks=5))
4
```

## System Info
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 8.0.61

OS: Debian GNU/Linux unstable (sid)
GCC version: (Debian 7.3.0-24) 7.3.0
CMake version: version 3.11.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.1.85
GPU models and configuration: GPU 0: GeForce GTX 950M
Nvidia driver version: 390.67
cuDNN version: Probably one of the following:
/usr/local/cuda/lib64/libcudnn.so.7.1.4
/usr/local/cuda/lib64/libcudnn_static.a

Versions of relevant libraries:
[pip3] numpy (1.14.5)
[pip3] torch (0.4.0)
[pip3] torchvision (0.2.1)
[conda] Could not collect

pytorch installed by pip3

cc @ezyang @gchanan @zou3519 @mruberry @SsnL

### PR DESCRIPTION
**Pull request by kurtamohler**
Fixes #9382

### CODE REVIEW

**Comment by kurtamohler** [comment]
There are at least two questions that need to be dealt with before we can merge this PR:
* Is `torch.tensor_split` an acceptable name, or is there a better name? Note Richard's points here: https://github.com/pytorch/pytorch/issues/9382#issuecomment-696735278
* Should we add a deprecation warning to `torch.chunk`, since this new function is meant to be a correction of its behavior? Or would there be any merit to keeping both functions?

**Comment by mruberry** [comment]
> There are at least two questions that need to be dealt with before we can merge this PR:
> 
> * Is `torch.tensor_split` an acceptable name, or is there a better name? Note Richard's points here: [#9382 (comment)](https://github.com/pytorch/pytorch/issues/9382#issuecomment-696735278)
> * Should we add a deprecation warning to `torch.chunk`, since this new function is meant to be an improvement upon its behavior? Or would there be any merit to keeping both functions?

Good questions.

`torch.tensor_split` seems like the appropriate name to me (and we can alias `torch.array_split` to it in the future). 

Let's not add the deprecation in this PR, but I agree we should consider it (for 1.8).

**Comment by kurtamohler** [comment]
I'm not sure why the JIT tests are failing on some machines and not others. They pass locally for me. I'll probably have to ssh into the jobs to debug.

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -1307,6 +1307,10 @@ def method_tests():
         ('split_with_sizes', (S, S, S), ([int(S / 3), S - int(S / 3) * 2, int(S / 3)],), '', (True,)),
         ('split_with_sizes', (S, S, S), ([int(S / 3), S - int(S / 3), 0],), 'size_0', (True, )),
         ('split_with_sizes', (S, S, S), ([int(S / 3), S - int(S / 3) * 2, int(S / 3)],), 'dim', (True, ), [1]),
+        ('tensor_split', (S, S, S), (3,), 'sections', (True,)),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:1310]: `(True,)` -> `(False,)` here and for the other entries

**Comment by mruberry** [comment]
> I'm not sure why the JIT tests are failing on some machines and not others. They pass locally for me (except for the new failure I introduced to `TestAutogradDeviceTypeCUDA.test_inplace_view_multi_output_safe_cuda`). I'll probably have to ssh into the jobs to debug.

It's because jit fusion is enabled on some builds and not others, I think, and method_tests is very hard to parse so it's hard to know what to set to tell the tests what kind of graphs are expected. I think if you just set `(True,)` to `(False,)` things will be fine.

In the future we'll use OpInfos, not method_tests(), and we won't have this problem.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (kurtamohler) [reply torch/testing/_internal/common_methods_invocations.py]: Thanks!

**Comment by mruberry** [comment]
The XLA variant of the test can be skipped for now by decorating it with `@onlyOnCPUAndCUDA`

**Comment by mruberry** [comment]
Hey @kurtamohler, is this ready for review?

**Comment by kurtamohler** [comment]
> Hey @kurtamohler, is this ready for review?

Yeah it is. I don't think my changes caused any of the CI failures.

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -1209,6 +1209,42 @@ def merge_dicts(*dicts):
 
 """.format(**common_args))
 
+add_docstr(torch.tensor_split,
+           r"""
+tensor_split(input, indices_or_sections, dim=0) -> List of Tensors
+
+Split a tensor into multiple sub-tensors, all of which are views of the
```
Comment (mruberry) [inline torch/_torch_docs.py:1216]: "Splits ... all of which are views of :attr:\`input`. This function is based on NumPy's array_split."

For the ellipses, can the one-sentence description incorporate the required indices_or_sections argument? Maybe something as simple as "into multiple subtensors based on the indices or number of sections specified by :attr:\`indices_or_sections\'"?

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -1209,6 +1209,42 @@ def merge_dicts(*dicts):
...
+input tensor. This function is based on :func:`numpy.array_split`.
+
+Args:
+    input (Tensor): the tensor to split
+    indices_or_sections (int or (list(int))):
+        If :attr:`indices_or_sections` is an integer, it represents the number of
```
Comment (mruberry) [inline torch/_torch_docs.py:1222]: Maybe something like:

"... is an integer n, :attr:\`input\` is split into n sections along dimension :attr:\`dim\`. If :attr:\`input\` is divisible by `n` along dimension :attr:\`\dim\`..."

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -1209,6 +1209,42 @@ def merge_dicts(*dicts):
...
+        sections to divide the input into. The input does not have to be divisible
+        by the number of sections. The first :code:`int(input.size(dim) % sections)`
+        output tensors will have have size :code:`int(input.size(dim) / sections) + 1`
+        in dimension :attr:`dim`. The rest will have size
+        :code:`int(input.size(dim) / sections)`.
+
+        If :attr:`indices_or_sections` is a list of ints, it represents a list of
```
Comment (mruberry) [inline torch/_torch_docs.py:1229]: This paragraph and the example seem great.

File: torch/_torch_docs.py [hunk 5]
```diff
@@ -1209,6 +1209,42 @@ def merge_dicts(*dicts):
...
+        indices where the input will be split. For instance, :code:`[2, 3]` and
```
Comment (mruberry) [inline torch/_torch_docs.py:1230]: "the input" -> ":attr:\`input\`"

File: torch/_torch_docs.py (hunk 4 above)
Comment (mruberry) [inline torch/_torch_docs.py:1229]: "It represents..." -> "then :attr:\`input\` is split along dimension :attr:\`dim\` at offsets..." ?

It's important to express that the splitting is along dim, and the docs should tie the list of ints directly to their effect instead of describing what they represent.

File: torch/_torch_docs.py [hunk 6]
```diff
@@ -1209,6 +1209,42 @@ def merge_dicts(*dicts):
...
+        :code:`dim=0` would result in the following tensors:
+
+            - :code:`input[:2]`
+            - :code:`input[2:3]`
+            - :code:`input[3:]`
+
+    dim (int, optional): dimension along which to split the tensor. Default: ``0``
+
+Example::
+    >>> x = torch.arange(8)
```
Comment (mruberry) [inline torch/_torch_docs.py:1240]: This is an excellent example.

File: torch/_torch_docs.py [hunk 7]
```diff
@@ -1209,6 +1209,42 @@ def merge_dicts(*dicts):
...
+    >>> torch.tensor_split(x, 3)
+    (tensor([0, 1, 2]), tensor([3, 4, 5]), tensor([6, 7]))
+    >>> x = torch.arange(7)
```
Comment (mruberry) [inline torch/_torch_docs.py:1243]: Also good. This needs an example where a list of integers is provided, and an example where dim != 0.

File: torch/_torch_docs.py [hunk 8]
```diff
@@ -1209,6 +1209,42 @@ def merge_dicts(*dicts):
 
 """.format(**common_args))
 
+add_docstr(torch.tensor_split,
+           r"""
+tensor_split(input, indices_or_sections, dim=0) -> List of Tensors
```
Comment (mruberry) [inline torch/_torch_docs.py:1214]: Reviewer's note: this is consistent with torch.chunk, which also does not support an out parameter and takes three positional args.

File: docs/source/torch.rst [hunk 9]
```diff
@@ -99,6 +99,7 @@ Indexing, Slicing, Joining, Mutating Ops
     stack
     t
     take
+    tensor_split
```
Comment (mruberry) [inline docs/source/torch.rst:102]: This also needs to update tensors.rst

File: aten/src/ATen/native/native_functions.yaml [hunk 10]
```diff
@@ -912,6 +912,16 @@
   variants: function, method
   device_guard: False
 
+- func: tensor_split.sections(Tensor(a) self, int sections, int dim=0) -> Tensor(a)[]
+  use_c10_dispatcher: full
+  variants: function, method
+  device_guard: False
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:918]: I wouldn't bother with device_guard: False here or below.

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 11]
```diff
@@ -172,6 +172,22 @@ std::vector<Tensor> chunk_batching_rule(const Tensor& self, int64_t chunks, int6
   return result;
 }
 
+std::vector<Tensor> tensor_split_sections_batching_rule(const Tensor& self, int64_t sections, int64_t dim) {
```
Comment (mruberry) [inline aten/src/ATen/BatchingRegistrations.cpp:175]: @zou3519 Do you want to take a look at these changes? Would you recommend PRs start updating this file?

File: aten/src/ATen/native/TensorShape.cpp [hunk 12]
```diff
@@ -513,6 +513,40 @@ std::vector<Tensor> chunk(const Tensor& self, int64_t chunks, int64_t dim) {
...
+static void check_tensor_split_dim(const Tensor& self, int64_t dim) {
+  TORCH_CHECK(self.dim() > 0, "expected at least a 1-dimensional tensor");
+  TORCH_CHECK(dim < self.dim(), "expected dim to be less than self.dim()");
+}
+
+std::vector<Tensor> tensor_split(const Tensor& self, int64_t sections, int64_t dim) {
+  check_tensor_split_dim(self, dim);
+  TORCH_CHECK(sections > 0, "number of sections must be larger than 0, got ", sections);
+  std::vector<Tensor> splits(sections);
+  int64_t min_split_size = self.size(dim) / sections;
```
Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:525]: Should this dim be wrapped before to use to support values like dim=-1?

File: aten/src/ATen/native/TensorShape.cpp [hunk 13]
```diff
@@ -513,6 +513,40 @@ std::vector<Tensor> chunk(const Tensor& self, int64_t chunks, int64_t dim) {
...
+}
+
+std::vector<Tensor> tensor_split(const Tensor& self, IntArrayRef indices, int64_t dim) {
+  check_tensor_split_dim(self, dim);
+  int64_t num_indices = indices.size();
+  std::vector<Tensor> splits(num_indices + 1);
+  int64_t start_idx = 0;
+  for (int64_t split_idx = 0; split_idx < num_indices; split_idx++) {
+    int64_t end_idx = indices[split_idx];
+    splits[split_idx] = at::slice(self, dim, start_idx, end_idx);
```
Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:543]: What happens when slicing past the end of an array? In NumPy, for example:

```
a = np.array((1, 2 ,3))
np.array_split(a, 6)
: [array([1]),
 array([2]),
 array([3]),
 array([], dtype=int64),
 array([], dtype=int64),
 array([], dtype=int64)]

np.array_split(a, (1, 2, 3, 7, 8))
: [array([1]),
 array([2]),
 array([3]),
 array([], dtype=int64),
 array([], dtype=int64),
 array([], dtype=int64)]
```

File: test/test_torch.py [hunk 14]
```diff
@@ -8334,6 +8334,100 @@ def test_contiguous(self, device):
...
+            (0,),
+            (10,),
+            (10, 0),
+            (0, 10),
+            (4, 10),
+            (12, 3),
+        ]
+        for input_size in input_sizes:
+            if dtype.is_floating_point or dtype.is_complex:
+                a = torch.randn(*input_size, dtype=dtype, device=device)
```
Comment (mruberry) [inline test/test_torch.py:8351]: We now have make_tensor in common_utils which might make this simpler

File: test/test_torch.py [hunk 15]
```diff
@@ -8334,6 +8334,100 @@ def test_contiguous(self, device):
         x.set_(x.storage(), 0, x.size(), stride)
         self.assertTrue(x.is_contiguous())
 
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.float, torch.double, torch.half, torch.int, torch.long, torch.cfloat, torch.cdouble, torch.bool, torch.uint8)
```
Comment (mruberry) [inline test/test_torch.py:8339]: Please sort these dtypes (uint8->complex128)

Comment (mruberry) [reply test/test_torch.py:8339]: We also have dtype acquisition helpers here:

https://github.com/pytorch/pytorch/blob/8a1e1004664e4cc8b1adbcef07b9fdae397e1787/torch/testing/__init__.py#L316

File: test/test_torch.py [hunk 16]
```diff
@@ -8334,6 +8334,100 @@ def test_contiguous(self, device):
...
+        for input_size in input_sizes:
+            if dtype.is_floating_point or dtype.is_complex:
+                a = torch.randn(*input_size, dtype=dtype, device=device)
+            else:
+                if dtype == torch.bool:
+                    a = torch.randint(2, input_size, dtype=dtype, device=device)
+                else:
+                    a = torch.randint(0, 127, input_size, dtype=dtype, device=device)
+            a_n = a.cpu().numpy()
+            for dim in range(a.dim()):
```
Comment (mruberry) [inline test/test_torch.py:8358]: This test looks pretty good. Would you try it with the tensor you construct and a transposed version of it, too, to test the operation on discontinguous tensors?

File: test/test_torch.py [hunk 17]
```diff
@@ -8334,6 +8334,100 @@ def test_contiguous(self, device):
...
+                for sections in range(1, 2 * a.size(dim)):
+                    msg = f'input_size {input_size}, sections {sections}, dim {dim}'
+                    result = torch.tensor_split(a, sections, dim)
+                    for result_item in result:
+                        self.assertEqual(result_item.device, torch.device(device), msg=msg)
```
Comment (mruberry) [inline test/test_torch.py:8363]: Great additional assertion

File: test/test_torch.py (hunk 16 above)
Comment (mruberry) [reply test/test_torch.py:8358]: Also let dim=-1

File: test/test_torch.py [hunk 18]
```diff
@@ -8334,6 +8334,100 @@ def test_contiguous(self, device):
...
+            (3,),
+            (10,),
+            (-1,),
+            (-10,),
+            (2, -1),
+            (3, 4, 10),
+            (0, -1, 0, 10),
+        ]
+        for input_size in input_sizes:
+            if dtype.is_floating_point or dtype.is_complex:
```
Comment (mruberry) [inline test/test_torch.py:8392]: make_tensor should help here, too

File: test/test_torch.py [hunk 19]
```diff
@@ -8334,6 +8334,100 @@ def test_contiguous(self, device):
...
+        indices_args = [
+            (),
+            (0,),
+            (3,),
+            (10,),
+            (-1,),
+            (-10,),
+            (2, -1),
+            (3, 4, 10),
+            (0, -1, 0, 10),
```
Comment (mruberry) [inline test/test_torch.py:8389]: What about one more like: (1, 5, 2, 8)?

File: test/test_torch.py [hunk 20]
```diff
@@ -8334,6 +8334,100 @@ def test_contiguous(self, device):
...
+        for input_size in input_sizes:
+            if dtype.is_floating_point or dtype.is_complex:
+                a = torch.randn(input_size, dtype=dtype, device=device)
+            else:
+                if dtype == torch.bool:
+                    a = torch.randint(2, input_size, dtype=dtype, device=device)
+                else:
+                    a = torch.randint(0, 127, input_size, dtype=dtype, device=device)
+            a_n = a.cpu().numpy()
+            for dim in range(a.dim()):
```
Comment (mruberry) [inline test/test_torch.py:8400]: dim=-1 here, too

File: test/test_torch.py [hunk 21]
```diff
@@ -8334,6 +8334,100 @@ def test_contiguous(self, device):
...
+        ]
+        for input_size in input_sizes:
+            if dtype.is_floating_point or dtype.is_complex:
+                a = torch.randn(input_size, dtype=dtype, device=device)
+            else:
+                if dtype == torch.bool:
+                    a = torch.randint(2, input_size, dtype=dtype, device=device)
+                else:
+                    a = torch.randint(0, 127, input_size, dtype=dtype, device=device)
+            a_n = a.cpu().numpy()
```
Comment (mruberry) [inline test/test_torch.py:8399]: as per above, also test on the transpose of the constructed tensor

File: test/test_torch.py [hunk 22]
```diff
@@ -8334,6 +8334,100 @@ def test_contiguous(self, device):
...
+            [(), (10,), 0, RuntimeError, r'expected at least a 1-dimensional tensor', IndexError],
+            [(S,), 0, 0, RuntimeError, r'number of sections must be larger than 0, got 0', ValueError],
+            [(S,), -1, 0, RuntimeError, r'number of sections must be larger than 0, got -1', ValueError],
+        ]
+        for input_size, sections_or_indices, dim, err, err_msg, numpy_err in test_cases:
+            a = torch.randn(input_size, device=device)
+            msg = f'input_size {input_size}, sections_or_indices {sections_or_indices}, dim {dim}'
+            with self.assertRaisesRegex(err, err_msg, msg=msg):
+                torch.tensor_split(a, sections_or_indices, dim)
+            if TEST_NUMPY:
```
Comment (mruberry) [inline test/test_torch.py:8427]: Good news: we actually import numpy always now, so we don't need these checks anymore

File: test/test_torch.py [hunk 23]
```diff
@@ -8334,6 +8334,100 @@ def test_contiguous(self, device):
...
+                for sections in range(1, 2 * a.size(dim)):
+                    msg = f'input_size {input_size}, sections {sections}, dim {dim}'
+                    result = torch.tensor_split(a, sections, dim)
+                    for result_item in result:
+                        self.assertEqual(result_item.device, torch.device(device), msg=msg)
+                        self.assertEqual(result_item.dtype, dtype, msg=msg)
+                    result_n = np.array_split(a_n, sections, dim)
+                    self.assertEqual(result_n, result, msg=msg)
+
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
```
Comment (mruberry) [inline test/test_torch.py:8368]: we can ditch this now since we always import NumPy

File: test/test_torch.py [hunk 24]
```diff
@@ -8334,6 +8334,100 @@ def test_contiguous(self, device):
         x.set_(x.storage(), 0, x.size(), stride)
         self.assertTrue(x.is_contiguous())
 
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
```
Comment (mruberry) [inline test/test_torch.py:8337]: we can ditch this now since we always import NumPy

File: test/test_torch.py [hunk 25]
```diff
@@ -8334,6 +8334,100 @@ def test_contiguous(self, device):
...
+                        self.assertEqual(result_item.dtype, dtype, msg=msg)
+                    result_n = np.array_split(a_n, indices, dim)
+                    self.assertEqual(result_n, result, msg=msg)
+
+    @onlyOnCPUAndCUDA
+    def test_tensor_split_errors(self, device):
+        S = 10
+        test_cases = [
+            # input size, sections or indices, dim, error type, error message, numpy error type
+            [(S,), 10, 1, RuntimeError, r'expected dim to be less than self.dim()', IndexError],
```
Comment (mruberry) [inline test/test_torch.py:8415]: I like how you formatted these test cases and then enumerated over them.

**Comment by mruberry** [comment]
Hey @kurtamohler! Another solid PR. I've asked @zou3519 to comment on one file and made a few suggestions on the docs and tests.

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 26]
```diff
@@ -527,6 +543,8 @@ TORCH_LIBRARY_IMPL(aten, Batched, m) {
 
   // view operations
   m.impl("chunk", chunk_batching_rule);
+  m.impl("tensor_split.sections", tensor_split_sections_batching_rule);
+  m.impl("tensor_split.indices", tensor_split_indices_batching_rule);
```
Comment (zou3519) [inline aten/src/ATen/BatchingRegistrations.cpp:547]: @kurtamohler could you add some tests to test_vmap.py similar to https://github.com/pytorch/pytorch/blob/67889db8aac73335ba01a2b0a01099e33e8c31a1/test/test_vmap.py#L1240-L1259?

`_vmap_view_test` eventually calls into https://github.com/pytorch/pytorch/blob/67889db8aac73335ba01a2b0a01099e33e8c31a1/test/test_vmap.py#L650-L655 , the comments there might be helpful. Please shout if you have questions!

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 11 above)
Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:175]: This looks correct; in another comment below I suggested adding a test.

I'd recommend against suggesting PR authors in general to update this file because the developer-facing API isn't stable and I haven't written any documentation. That being said, contributions are appreciated and the below code looks good to me.

File: aten/src/ATen/native/TensorShape.cpp (hunk 12 above)
Comment (kurtamohler) [reply aten/src/ATen/native/TensorShape.cpp:525]: Good idea. The `maybe_wrap_dim` call will perform error checking, so I can remove the `dim < self.dim()` check from above.

File: aten/src/ATen/native/TensorShape.cpp (hunk 13 above)
Comment (kurtamohler) [reply aten/src/ATen/native/TensorShape.cpp:543]: Yeah, `at::slice` works correctly in this case:
```
>>> torch.tensor_split(torch.tensor([1, 2, 3]), 6)
(tensor([1]), tensor([2]), tensor([3]), tensor([], dtype=torch.int64), tensor([], dtype=torch.int64), tensor([], dtype=torch.int64))
>>> torch.tensor_split(torch.tensor([1, 2, 3]), (1, 2, 3, 7, 8))
(tensor([1]), tensor([2]), tensor([3]), tensor([], dtype=torch.int64), tensor([], dtype=torch.int64), tensor([], dtype=torch.int64))
```

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 26 above)
Comment (kurtamohler) [reply aten/src/ATen/BatchingRegistrations.cpp:547]: Done! I used pretty similar test cases as `test_split`, but with a few differences.

File: torch/_torch_docs.py (hunk 7 above)
Comment (kurtamohler) [reply torch/_torch_docs.py:1247]: Good idea, added.

File: torch/_torch_docs.py (hunk 2 above)
Comment (kurtamohler) [reply torch/_torch_docs.py]: Done

File: test/test_torch.py (hunk 15 above)
Comment (kurtamohler) [reply test/test_torch.py]: Changed to `get_all_dtypes`. Just FYI, I'm skipping BFloat because numpy doesn't support it.

File: test/test_torch.py (hunk 20 above)
Comment (kurtamohler) [reply test/test_torch.py]: I decided to go all the way down to `-a.dim()` just to be sure.

File: test/test_torch.py (hunk 15 above)
Comment (mruberry) [reply test/test_torch.py]: That's OK. Would you just add a brief comment as to why bfloat16 is skipped at the top of the test?

File: torch/_torch_docs.py (hunk 3 above)
Comment (kurtamohler) [reply torch/_torch_docs.py]: Done

**Comment by kurtamohler** [comment]
Thanks for the comments @mruberry  and @zou3519! Latest update should cover everything brought up so far.

File: test/test_torch.py (hunk 15 above)
Comment (kurtamohler) [reply test/test_torch.py]: Done

**Review by mruberry — APPROVED** [review]
Looks great to me!

@zou3519 would you like to review the updates to the vmap-related code?

**Review by zou3519 — APPROVED** [review]
vmap related parts lgtm
